worker/provisioner/provisioner_task.go

```diff
@@ -755,35 +779,41 @@ func (task *provisionerTask) startMachine(
 	provisioningInfo *params.ProvisioningInfo,
 	startInstanceParams environs.StartInstanceParams,
 ) error {
-	var result *environs.StartInstanceResult
 	// TODO (jam): 2017-01-19 Should we be setting this earlier in the cycle?
 	if err := machine.SetInstanceStatus(status.Provisioning, "starting", nil); err != nil {
 		logger.Errorf("%v", err)
 	}
-	for attemptsLeft := task.retryStartInstanceStrategy.retryCount; attemptsLeft >= 0; attemptsLeft-- {
-		attemptResult, err := task.broker.StartInstance(startInstanceParams)
-		if err == nil {
-			result = attemptResult
-			break
-		} else if attemptsLeft <= 0 {
-			// Set the state to error, so the machine will be skipped
-			// next time until the error is resolved, but don't return
-			// an error; just keep going with the other machines.
+	var result *environs.StartInstanceResult
+	err := retry.Call(retry.CallArgs{
+		Attempts: task.retryStartInstanceStrategy.retryCount,
+		Clock:    task.clock,
+		Delay:    task.retryStartInstanceStrategy.retryDelay,
+		Func: func() error {
+			attemptResult, err := task.broker.StartInstance(startInstanceParams)
+			if err == nil {
+				result = attemptResult
+				return nil
+			}
+			if task.catacomb.Dying() {
+				return task.catacomb.ErrDying()
+			}
 			return task.setErrorStatus("cannot start instance for machine %q: %v", machine, err)
-		}
-
-		retryMsg := fmt.Sprintf("failed to start instance (%s), retrying in %v (%d more attempts)",
-			err.Error(), task.retryStartInstanceStrategy.retryDelay, attemptsLeft)
-		logger.Warningf(retryMsg)
-		if err2 := machine.SetInstanceStatus(status.Provisioning, retryMsg, nil); err2 != nil {
-			logger.Errorf("%v", err2)
-		}
-
-		select {
-		case <-task.catacomb.Dying():
-			return task.catacomb.ErrDying()
-		case <-time.After(task.retryStartInstanceStrategy.retryDelay):
-		}
+		},
+		//BackoffFunc func(delay time.Duration, attempt int) time.Duration
+		NotifyFunc: func(lastError error, attempt int) {
+			retryMsg := fmt.Sprintf("failed to start instance (%s), retrying in %v (%d more attempts)",
+				lastError.Error(), task.retryStartInstanceStrategy.retryDelay, task.retryStartInstanceStrategy.retryCount-attempt)
+			logger.Warningf(retryMsg)
+			if err2 := machine.SetInstanceStatus(status.Provisioning, retryMsg, nil); err2 != nil {
+				logger.Errorf("%v", err2)
+			}
+		},
+		IsFatalError: func(err error) bool {
+			return fmt.Sprintf("catacomb %p is dying", task.catacomb) == err.Error()
```

Comment:
Doing a string comparison here breaks encapsulation quite badly. Can't we just return true here?